fix: file write race condition — proven data loss on concurrent appends

## Problem

`append_to_note` (and all file write operations in `write.py`, `edit.py`, `structure.py`, `inbox.py`) perform a non-atomic read-modify-write cycle without any file locking. Two concurrent writers to the same file will lose one write.

**This is a proven bug** — `test_concurrent_append_no_data_loss` in `tests/unit/test_concurrency.py` fails consistently, demonstrating the race:

```
FAILED tests/unit/test_concurrency.py::TestWriteConcurrency::test_concurrent_append_no_data_loss
AssertionError: assert 'THREAD_B_CONTENT' in <file that only contains THREAD_A_CONTENT>
```

The race is exploitable in production because:
- The watcher's ingest thread runs `ingest()` which may write files
- The MCP tool handler calls `append_to_note`, `create_note`, etc.
- The event listener in `server.py` processes note change events
- All three can touch the same file simultaneously

## Location

- `src/alaya/tools/write.py` — `append_to_note()` (L88), `create_note()` (L44), `update_tags()` (L138)
- `src/alaya/tools/edit.py` — `replace_section()` (L31)
- `src/alaya/tools/inbox.py` — `capture_to_inbox()` (L15), `clear_inbox_item()` (L44)
- `src/alaya/tools/structure.py` — `rename_note()` (L95), `delete_note()` (L127)

## Proposed Fix

Add a per-path `threading.Lock`. Keep a `dict[Path, Lock]` in a small `_locks.py` module. Wrap all read-modify-write cycles:

```python
# src/alaya/tools/_locks.py
import threading
from pathlib import Path

_path_locks: dict[Path, threading.Lock] = {}
_registry_lock = threading.Lock()

def get_path_lock(path: Path) -> threading.Lock:
    resolved = path.resolve()
    with _registry_lock:
        if resolved not in _path_locks:
            _path_locks[resolved] = threading.Lock()
        return _path_locks[resolved]
```

Then: `with get_path_lock(path): read -> modify -> write`

No new deps. In-process only (MCP is single-process).

## Definition of Done

- [ ] `_locks.py` module with per-path lock registry
- [ ] All file read-modify-write operations wrapped in path lock
- [ ] `test_concurrent_append_no_data_loss` passes 100/100
- [ ] New tests: concurrent `update_tags`, concurrent `replace_section` on same file
- [ ] No deadlock across different files

## PRD

**Goal:** Eliminate data loss from concurrent file writes.
**Scope:** All functions in `write.py`, `edit.py`, `inbox.py`, `structure.py` that read-modify-write vault files.
**Non-goals:** Cross-process locking; LanceDB locking (handled internally).
**Acceptance:** Existing + new concurrency tests pass consistently. No new dependencies.

--- a/src/alaya/tools/_locks.py
+++ b/src/alaya/tools/_locks.py
@@ -1,0 +1,41 @@
+"""Per-path file locks for atomic read-modify-write operations.
+
+All vault file writes acquire a per-path lock so concurrent MCP tool calls
+(FastMCP dispatches sync tools to a thread pool) cannot interleave reads and
+writes on the same file.
+
+Only protects within a single process — sufficient for the MCP server model.
+"""
+import threading
+from pathlib import Path
+
+_registry_lock = threading.Lock()
+_path_locks: dict[Path, threading.Lock] = {}
+
+
+def get_path_lock(path: Path) -> threading.Lock:
+    """Return the lock for *path*, creating it if needed.
+
+    Uses path.resolve() so that relative and absolute references to the
+    same file share a single lock.
+    """
+    resolved = path.resolve()
+    with _registry_lock:
+        if resolved not in _path_locks:
+            _path_locks[resolved] = threading.Lock()
+        return _path_locks[resolved]
+
+
+def atomic_write(path: Path, content: str) -> None:
+    """Write *content* to *path* atomically via a sibling temp file.
+
+    On POSIX, os.replace() is guaranteed atomic when src and dst are on the
+    same filesystem (which is always true for a sibling temp file).
+    """
+    tmp = path.with_suffix(".tmp")
+    try:
+        tmp.write_text(content)
+        tmp.replace(path)
+    except Exception:
+        tmp.unlink(missing_ok=True)
+        raise

--- a/src/alaya/tools/edit.py
+++ b/src/alaya/tools/edit.py
@@ -5,6 +5,7 @@ from fastmcp import FastMCP
 from alaya.errors import error, NOT_FOUND, OUTSIDE_VAULT, SECTION_NOT_FOUND, ALREADY_EXISTS
 from alaya.vault import resolve_note_path
 from alaya.tools.write import create_note
+from alaya.tools._locks import get_path_lock, atomic_write
 
 
 def _parse_sections(content: str) -> list[tuple[str, int, int]]:
@@ -39,26 +40,28 @@ def replace_section(
     Raises ValueError with 'SECTION_NOT_FOUND' if the section header doesn't exist.
     """
     path = resolve_note_path(relative_path, vault)
-    if not path.exists():
-        raise FileNotFoundError(f"Note not found: {relative_path}")
 
-    content = path.read_text()
-    lines = content.splitlines()
-    sections = _parse_sections(content)
+    with get_path_lock(path):
+        if not path.exists():
+            raise FileNotFoundError(f"Note not found: {relative_path}")
 
-    match = next((s for s in sections if s[0].lower() == section.lower()), None)
-    if match is None:
-        raise ValueError(f"SECTION_NOT_FOUND: '{section}' in {relative_path}")
+        content = path.read_text()
+        lines = content.splitlines()
+        sections = _parse_sections(content)
 
-    header_text, start, end = match
+        match = next((s for s in sections if s[0].lower() == section.lower()), None)
+        if match is None:
+            raise ValueError(f"SECTION_NOT_FOUND: '{section}' in {relative_path}")
 
-    # rebuild: everything before section body + header + new content + rest
-    header_line = lines[start]
-    before = lines[:start]
-    after = lines[end:]
+        header_text, start, end = match
 
-    new_lines = before + [header_line, new_content] + ([""] if after else []) + after
-    path.write_text("\n".join(new_lines) + "\n")
+        # rebuild: everything before section body + header + new content + rest
+        header_line = lines[start]
+        before = lines[:start]
+        after = lines[end:]
+
+        new_lines = before + [header_line, new_content] + ([""] if after else []) + after
+        atomic_write(path, "\n".join(new_lines) + "\n")
 
 
 def extract_section(

--- a/src/alaya/tools/inbox.py
+++ b/src/alaya/tools/inbox.py
@@ -4,6 +4,7 @@ from pathlib import Path
 
 from fastmcp import FastMCP
 from alaya.errors import error, NOT_FOUND
+from alaya.tools._locks import get_path_lock, atomic_write
 
 _INBOX_FILENAME = "inbox.md"
 
@@ -18,9 +19,10 @@ def capture_to_inbox(text: str, vault: Path) -> str:
     timestamp = datetime.now().strftime("%Y-%m-%d %H:%M")
     entry = f"- {timestamp} {text}"
 
-    existing = inbox.read_text() if inbox.exists() else "# Inbox\n\nQuick capture. Process weekly.\n"
-    separator = "\n" if existing.endswith("\n") else "\n"
-    inbox.write_text(existing + separator + entry + "\n")
+    with get_path_lock(inbox):
+        existing = inbox.read_text() if inbox.exists() else "# Inbox\n\nQuick capture. Process weekly.\n"
+        separator = "\n" if existing.endswith("\n") else "\n"
+        atomic_write(inbox, existing + separator + entry + "\n")
 
     return f"Captured: {entry}"
 
@@ -53,24 +55,25 @@ def clear_inbox_item(text: str, vault: Path) -> None:
     """
     import re
     inbox = _inbox_path(vault)
-    lines = inbox.read_text().splitlines(keepends=True)
-
     _TS_PREFIX = re.compile(r"^-\s+\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}\s+")
 
-    # first pass: exact content match (strip timestamp prefix)
-    for i, line in enumerate(lines):
-        stripped = _TS_PREFIX.sub("", line.rstrip())
-        if stripped == text:
-            lines.pop(i)
-            inbox.write_text("".join(lines))
-            return
+    with get_path_lock(inbox):
+        lines = inbox.read_text().splitlines(keepends=True)
 
-    # second pass: substring match fallback
-    for i, line in enumerate(lines):
-        if text in line:
-            lines.pop(i)
-            inbox.write_text("".join(lines))
-            return
+        # first pass: exact content match (strip timestamp prefix)
+        for i, line in enumerate(lines):
+            stripped = _TS_PREFIX.sub("", line.rstrip())
+            if stripped == text:
+                lines.pop(i)
+                atomic_write(inbox, "".join(lines))
+                return
+
+        # second pass: substring match fallback
+        for i, line in enumerate(lines):
+            if text in line:
+                lines.pop(i)
+                atomic_write(inbox, "".join(lines))
+                return
 
     raise ValueError(f"Inbox item not found: '{text}'")
 

--- a/src/alaya/tools/structure.py
+++ b/src/alaya/tools/structure.py
@@ -8,6 +8,7 @@ from alaya.errors import error, NOT_FOUND, OUTSIDE_VAULT, INVALID_ARGUMENT
 from alaya.events import emit, NoteEvent
 from alaya.vault import resolve_note_path
 from alaya.tools.write import _validate_directory, _slugify
+from alaya.tools._locks import get_path_lock, atomic_write
 
 _ARCHIVES_DIR = "archives"
 
@@ -103,18 +104,17 @@ def rename_note(relative_path: str, new_title: str, vault: Path) -> str:
 
     # Use frontmatter title as the wikilink key; fall back to stem when absent.
     # zk wikilinks reference the note title, not the filename.
-    content = src.read_text()
-    fm_title_match = re.search(r"^title:\s*(.+)$", content, re.MULTILINE)
-    old_title = fm_title_match.group(1).strip() if fm_title_match else src.stem
-    new_slug = _slugify(new_title)
-    dest = src.parent / f"{new_slug}.md"
+    with get_path_lock(src):
+        content = src.read_text()
+        fm_title_match = re.search(r"^title:\s*(.+)$", content, re.MULTILINE)
+        old_title = fm_title_match.group(1).strip() if fm_title_match else src.stem
+        new_slug = _slugify(new_title)
+        dest = src.parent / f"{new_slug}.md"
 
-    # update frontmatter title (reuse already-read content)
-    content = re.sub(r"^title:.*$", f"title: {new_title}", content, count=1, flags=re.MULTILINE)
-    src.write_text(content)
-
-    # rename file
-    src.rename(dest)
+        # update frontmatter title then rename atomically
+        content = re.sub(r"^title:.*$", f"title: {new_title}", content, count=1, flags=re.MULTILINE)
+        atomic_write(src, content)
+        src.rename(dest)
 
     # update all [[old_title]] → [[new_slug]] across vault
     find_and_replace_wikilinks(old_title, new_slug, vault)
@@ -142,8 +142,9 @@ def delete_note(relative_path: str, vault: Path, reason: str | None = None) -> s
             raise
 
     if reason:
-        existing = src.read_text()
-        src.write_text(_insert_frontmatter_field(existing, "archived_reason", reason))
+        with get_path_lock(src):
+            existing = src.read_text()
+            atomic_write(src, _insert_frontmatter_field(existing, "archived_reason", reason))
 
     archives_dir = vault / _ARCHIVES_DIR
     archives_dir.mkdir(exist_ok=True)

--- a/src/alaya/tools/write.py
+++ b/src/alaya/tools/write.py
@@ -8,6 +8,7 @@ from fastmcp import FastMCP
 from alaya.errors import error, NOT_FOUND, ALREADY_EXISTS, OUTSIDE_VAULT, INVALID_ARGUMENT
 from alaya.events import emit, NoteEvent
 from alaya.vault import resolve_note_path
+from alaya.tools._locks import get_path_lock, atomic_write
 
 # Directories considered valid targets for note creation
 _VALID_DIRS = {
@@ -76,10 +77,11 @@ def create_note(
     if body:
         content_parts.append(body)
 
-    if file_path.exists():
-        raise FileExistsError(f"Note already exists: {file_path.relative_to(vault)}")
+    with get_path_lock(file_path):
+        if file_path.exists():
+            raise FileExistsError(f"Note already exists: {file_path.relative_to(vault)}")
+        atomic_write(file_path, "\n".join(content_parts) + "\n")
 
-    file_path.write_text("\n".join(content_parts) + "\n")
     relative = str(file_path.relative_to(vault))
     emit(NoteEvent("created", relative))
     return relative
@@ -98,40 +100,43 @@ def append_to_note(
     dated: prepend a '### YYYY-MM-DD' heading to the appended text.
     """
     path = resolve_note_path(relative_path, vault)
-    if not path.exists():
-        raise FileNotFoundError(f"Note not found: {relative_path}")
 
     if dated:
         text = f"### {date.today().isoformat()}\n{text}"
 
-    existing = path.read_text()
+    with get_path_lock(path):
+        if not path.exists():
+            raise FileNotFoundError(f"Note not found: {relative_path}")
 
-    if section_header is None:
-        separator = "\n" if existing.endswith("\n") else "\n\n"
-        path.write_text(existing + separator + text + "\n")
-        emit(NoteEvent("modified", relative_path))
-        return
+        existing = path.read_text()
 
-    # Insert under the named section, before the next ## heading or EOF
-    lines = existing.splitlines(keepends=True)
-    target = f"## {section_header}"
-    section_idx = next(
-        (i for i, l in enumerate(lines) if l.rstrip() == target),
-        None,
-    )
-    if section_idx is None:
-        raise ValueError(f"Section not found: '{section_header}'")
+        if section_header is None:
+            separator = "\n" if existing.endswith("\n") else "\n\n"
+            atomic_write(path, existing + separator + text + "\n")
+            emit(NoteEvent("modified", relative_path))
+            return
 
-    # find insertion point: end of this section (before the next ## or EOF)
-    insert_at = len(lines)
-    for i in range(section_idx + 1, len(lines)):
-        if lines[i].startswith("## "):
-            insert_at = i
-            break
+        # Insert under the named section, before the next ## heading or EOF
+        lines = existing.splitlines(keepends=True)
+        target = f"## {section_header}"
+        section_idx = next(
+            (i for i, l in enumerate(lines) if l.rstrip() == target),
+            None,
+        )
+        if section_idx is None:
+            raise ValueError(f"Section not found: '{section_header}'")
 
-    # inject a blank line + text before the insertion point
-    new_lines = lines[:insert_at] + ["\n", text + "\n"] + lines[insert_at:]
-    path.write_text("".join(new_lines))
+        # find insertion point: end of this section (before the next ## or EOF)
+        insert_at = len(lines)
+        for i in range(section_idx + 1, len(lines)):
+            if lines[i].startswith("## "):
+                insert_at = i
+                break
+
+        # inject a blank line + text before the insertion point
+        new_lines = lines[:insert_at] + ["\n", text + "\n"] + lines[insert_at:]
+        atomic_write(path, "".join(new_lines))
+
     emit(NoteEvent("modified", relative_path))
 
 
@@ -142,58 +147,57 @@ def update_tags(relative_path: str, add: list[str], remove: list[str], vault: Pa
     If the tag line doesn't exist, a new one is created after the closing ---.
     """
     path = resolve_note_path(relative_path, vault)
-    if not path.exists():
-        raise FileNotFoundError(f"Note not found: {relative_path}")
 
-    content = path.read_text()
+    with get_path_lock(path):
+        if not path.exists():
+            raise FileNotFoundError(f"Note not found: {relative_path}")
 
-    # find existing inline tag line (first non-empty line after closing ---)
-    lines = content.splitlines()
-    fm_end = -1
-    in_fm = False
-    for i, line in enumerate(lines):
-        if line.strip() == "---":
-            if not in_fm:
-                in_fm = True
-            else:
-                fm_end = i
-                break
+        content = path.read_text()
 
-    if fm_end == -1:
-        # no frontmatter, treat the whole file
+        # find existing inline tag line (first non-empty line after closing ---)
+        lines = content.splitlines()
         fm_end = -1
+        in_fm = False
+        for i, line in enumerate(lines):
+            if line.strip() == "---":
+                if not in_fm:
+                    in_fm = True
+                else:
+                    fm_end = i
+                    break
 
-    # find the tag line: first line after frontmatter that contains only #words
-    tag_line_idx = None
-    for i in range(fm_end + 1, len(lines)):
-        line = lines[i].strip()
-        if not line:
-            continue
-        if re.match(r"^(#\w[\w-]* ?)+$", line):
-            tag_line_idx = i
-        break
+        # find the tag line: first line after frontmatter that contains only #words
+        tag_line_idx = None
+        for i in range(fm_end + 1, len(lines)):
+            line = lines[i].strip()
+            if not line:
+                continue
+            if re.match(r"^(#\w[\w-]* ?)+$", line):
+                tag_line_idx = i
+            break
 
-    if tag_line_idx is not None:
-        existing_tags = set(re.findall(r"#([\w-]+)", lines[tag_line_idx]))
-    else:
-        existing_tags = set()
+        if tag_line_idx is not None:
+            existing_tags = set(re.findall(r"#([\w-]+)", lines[tag_line_idx]))
+        else:
+            existing_tags = set()
 
-    updated_tags = (existing_tags | set(add)) - set(remove)
+        updated_tags = (existing_tags | set(add)) - set(remove)
 
-    # nothing changed — skip the write to avoid reordering tags
-    if updated_tags == existing_tags and not add:
-        return
+        # nothing changed — skip the write to avoid reordering tags
+        if updated_tags == existing_tags and not add:
+            return
 
-    new_tag_line = " ".join(f"#{t}" for t in sorted(updated_tags))
+        new_tag_line = " ".join(f"#{t}" for t in sorted(updated_tags))
 
-    if tag_line_idx is not None:
-        lines[tag_line_idx] = new_tag_line
-    else:
-        # insert after frontmatter closing ---
-        insert_at = fm_end + 1 if fm_end >= 0 else 0
-        lines.insert(insert_at, new_tag_line)
+        if tag_line_idx is not None:
+            lines[tag_line_idx] = new_tag_line
+        else:
+            # insert after frontmatter closing ---
+            insert_at = fm_end + 1 if fm_end >= 0 else 0
+            lines.insert(insert_at, new_tag_line)
 
-    path.write_text("\n".join(lines) + "\n")
+        atomic_write(path, "\n".join(lines) + "\n")
+
     emit(NoteEvent("modified", relative_path))
 
 

--- a/tests/unit/test_concurrency.py
+++ b/tests/unit/test_concurrency.py
@@ -144,3 +144,60 @@ class TestWriteConcurrency:
         assert len(results) == 2
         assert (vault / results[0]).exists()
         assert (vault / results[1]).exists()
+
+    def test_concurrent_update_tags_no_data_loss(self, vault: Path) -> None:
+        """Two threads adding different tags must both survive in the file."""
+        from alaya.tools.write import update_tags
+
+        note_path = "projects/second-brain.md"
+        barrier = threading.Barrier(2)
+        errors = []
+
+        def add_tag(tag):
+            try:
+                barrier.wait()
+                update_tags(note_path, add=[tag], remove=[], vault=vault)
+            except Exception as e:
+                errors.append(e)
+
+        t1 = threading.Thread(target=add_tag, args=("tag-alpha",))
+        t2 = threading.Thread(target=add_tag, args=("tag-beta",))
+        t1.start(); t2.start()
+        t1.join(); t2.join()
+
+        assert not errors, f"Concurrent update_tags raised: {errors}"
+        content = (vault / note_path).read_text()
+        assert "tag-alpha" in content
+        assert "tag-beta" in content
+
+    def test_concurrent_replace_section_no_data_loss(self, vault: Path) -> None:
+        """Two threads replacing different sections must both succeed."""
+        from alaya.tools.edit import replace_section
+
+        note_path = "projects/second-brain.md"
+        # ensure two sections exist
+        (vault / note_path).write_text(
+            "---\ntitle: Test\ndate: 2026-01-01\n---\n"
+            "## Alpha\noriginal alpha\n"
+            "## Beta\noriginal beta\n"
+        )
+
+        barrier = threading.Barrier(2)
+        errors = []
+
+        def replace(section, new_content):
+            try:
+                barrier.wait()
+                replace_section(note_path, section, new_content, vault)
+            except Exception as e:
+                errors.append(e)
+
+        t1 = threading.Thread(target=replace, args=("Alpha", "new alpha content"))
+        t2 = threading.Thread(target=replace, args=("Beta", "new beta content"))
+        t1.start(); t2.start()
+        t1.join(); t2.join()
+
+        assert not errors, f"Concurrent replace_section raised: {errors}"
+        content = (vault / note_path).read_text()
+        assert "new alpha content" in content
+        assert "new beta content" in content

